storages/shardclient.go

```diff
@@ -0,0 +1,89 @@
+package storages
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/allegro/akubra/balancing"
+	"github.com/allegro/akubra/log"
+
+	set "github.com/deckarep/golang-set"
+)
+
+// NamedShardClient interface
+type NamedShardClient interface {
+	http.RoundTripper
+	Name() string
+	Backends() []*StorageClient
+}
+
+// ShardClient stores information about cluster backends
+type ShardClient struct {
+	backends          []*StorageClient
+	name              string
+	synclog           *SyncSender
+	MethodSet         set.Set
+	requestDispatcher dispatcher
+	balancer          *balancing.BalancerPrioritySet
+}
+
+// RoundTrip implements http.RoundTripper interface
+func (c *ShardClient) RoundTrip(req *http.Request) (*http.Response, error) {
+
+	reqID, _ := req.Context().Value(log.ContextreqIDKey).(string)
+	log.Debugf("Shard: Got request id %s", reqID)
+	if c.balancer != nil && (req.Method == http.MethodGet || req.Method == http.MethodHead || req.Method == http.MethodOptions) {
+		resp, err := c.balancerRoundTrip(req)
+		log.Debugf("Request %s, processed by balancer error %s", reqID, err)
+		return resp, err
+
+	}
+	log.Debug("Request %s processed by dispatcher, reqId")
+	return c.requestDispatcher.Dispatch(req)
+}
+
+func (c *ShardClient) balancerRoundTrip(req *http.Request) (resp *http.Response, err error) {
+	notFoundNodes := []balancing.Node{}
```

Comment:
maybe its worth it to add entries to sync log for each of the nodes in `notFoundNodes` after we've found a node that was able to process the request?